src/mlpack/methods/ann/init_rules/glorot_init.hpp

```diff
@@ -0,0 +1,126 @@
+/**
+ * @file glorot_init.hpp
+ * @author Prabhat Sharma
+ *
+ * Definition and implementation of the Glorot initialization method. This
+ * initialization rule initialize the weights to maintain activation variances
+ * and back-propagated gradients variance as one moves up or down the network.
+ *
+ * For more information, see the following paper.
+ *
+ * @code
+ * @inproceedings{pmlr-v9-glorot10a,
+ * title={Understanding the difficulty of training deep feedforward neural networks},
+ * author={Xavier Glorot and Yoshua Bengio},
+ * booktitle={Proceedings of the Thirteenth International Conference on Artificial
+ *              Intelligence and Statistics},
+ * year={2010}
+ * }
+ * @endcode
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_INIT_RULES_GLOROT_INIT_HPP
+#define MLPACK_METHODS_ANN_INIT_RULES_GLOROT_INIT_HPP
+
+#include <mlpack/prereqs.hpp>
+#include "random_init.hpp"
+#include "gaussian_init.hpp"
+
+using namespace mlpack::math;
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * This class is used to initialize the weight matrix with the Glorot Initialization
+ * method. The method is defined by
+ *
+ * @f{eqnarray*}{
+ * \Var[w_i] &=& \frac{2}{n_i + n_{i+1}} \\
+ * w_i ~ \U{-\frac{\sqrt(6)}{\sqrt(n_i + n_{i+1})}, \frac{\sqrt(6)}{\sqrt(n_i + n_{i+1})}}
+ * @f}
+ * Where n_{i+1} is the number of neurons in the outgoing layer, n_i represents the
+ * number of neurons in the ingoing layer
+ * Here Normal Distribution may also be used if needed
+ */
+template<bool Uniform = true>
+class GlorotInitializationType
+{
+ public:
+  /**
+   * Initialize
+   */
+  GlorotInitializationType()
+  {
+    // Nothing to do here.
+  }
+
+  /**
+   * Initialize the elements weight matrix.
+   *
+   * @param W Weight matrix to initialize.
+   * @param rows Number of rows.
+   * @param cols Number of columns.
+   */
+  template<typename eT>
+  void Initialize(arma::Mat<eT>& W,
```

Comment:
Do you think we could do something like:

```c++
GlorotInitializationType<true>::Initialize(..)
{
  double_t a = sqrt(6)/sqrt(rows + cols); // limit of  distribution
  RandomInitialization randomInit(-a, a);
  randomInit.Initialize(W, rows, cols);
}
```

```c++
GlorotInitializationType<false>::Initialize(..)
{
  double_t var = (double)(2)/(rows + cols);
  GaussianInitialization normalInit(0.0, var);
  normalInit.Initialize(W, rows, cols);
}
```

Reply:
Done :)